### PR TITLE
fix(extract): preserve executable perms on extracted files

### PR DESCRIPTION
**Pull request by zkat**
This is basically a copy-paste from the old way npm
would set permissions on extract, and should now be
fully compatible again. I hope. Sigh.

### CODE REVIEW

File: lib/extract-stream.js
Comment (mikesherov) [inline lib/extract-stream.js:21]: Came here to say that `~(opts.umask || 0)` isn't the same as `(~opts.umask || 0)`, but for all intents and purposes, it is 0_o

**Review by mikesherov — APPROVED** [review]
So is the difference here that fmode and fmode are additive privs rather than replacement private? Looks great! 

Perhaps consider adding a unit test considering this was a regression.

**Comment by iarna** [comment]
@zkat's conferencing so I intend to extract this function and write detailed unit tests.

**Comment by mikesherov** [comment]
Cool. Then maybe we land it for now so we can unbreak npm?

**Comment by iarna** [comment]
@mikesherov I'm gonna do the test first, but planning it for today. I have _never_ seen a test written for something after it lands. _ever_. =p

**Comment by mikesherov** [comment]
Fair enough!

**Comment by rmg** [comment]
The piece that was missing from the existing tests is giving a `mode` for each entry to `mockTar`. The current tests (and behaviour before this PR) appear to ignore per-entry permissions entirely and instead assume some sort of archive-wide dmode/fmode, which isn't actually a thing.

**Comment by benjamn** [comment]
For what it's worth, I can confirm that this PR fixes the problem for Meteor (see my [commit](https://github.com/meteor/meteor/commit/3431b5b81f1477165c053f6816901b1d5b634acd) linked above). Thanks! 🙌 

Note to other project/package maintainers: it was possible for Meteor to pull in this change early because we bundle a specific version of `npm`, and thus have the ability to override the version of `pacote` that our `npm` uses. I would not expect this approach to work for everyone, and I look forward to reverting my commit when `pacote` is released again.

We do have a [test](https://github.com/meteor/meteor/blob/5f13b20e1e4023a23a50b9c4f1c9e1ad9d88136c/tools/tests/npm.js#L17-L20) for executable bit preservation in npm packages, which was failing with `npm@5.4.0` and `pacote@6.0.1`, and now passes.

**Review by benjamn — APPROVED** [review]

**Comment by rmg** [comment]
@zkat @iarna I've submitted #118 as a PR against this branch to clean up the tests.

**Comment by iarna** [comment]
This is landed in Pacote 6.0.2 and will be an new npm soon.

**Pull request closed without merging**
